THE ASSIGNMENT IS ALL ABOUT LOADING THE DATASET AND PERFORM EDA PROCESS ON TOP OF THE DATASET WHICH IS IN PARQUET FILE FORMAT.


In [1]:
%pip install numpy==1.23.5
%pip install pandas==1.5.3
%pip install sweetviz==2.1
%pip install pyarrow


Defaulting to user installation because normal site-packages is not writeable
Note: you may need to restart the kernel to use updated packages.
Defaulting to user installation because normal site-packages is not writeable
Note: you may need to restart the kernel to use updated packages.
Defaulting to user installation because normal site-packages is not writeable
Note: you may need to restart the kernel to use updated packages.
Defaulting to user installation because normal site-packages is not writeable
Note: you may need to restart the kernel to use updated packages.


In [2]:
import pandas as pd 
import numpy as np
import os
# import pyarrow

Loading data as parquet files

In [3]:
source_path="/home/ai/Documents/my files/assignment_2/Pyspark_Assignment_2/raw_data_in/hotelsourcesdetails.parquet"

In [4]:
hotel_details=pd.read_parquet(source_path)
hotel_details.head(5)

id  hotelcode   hotelid  websitecode                    hotelname  \
0  4492402     329109  22995521            1                Hotel Perales   
1  4492405     329778   1381371            1  The Cherry Tree Guest House   
2  4492408     333046  17810920            1            Asia Boutique Inn   
3  4492409     334152  46258460            1          Dunham's Bay Resort   
4  4492410     334788  31836214            1          Pazo de Turbisquedo   

                                             address                  city  \
0                                 Avenida Pio XII, 3  Talavera de la Reina   
1                                 15 Silverdale Road            Eastbourne   
2  Rua Correia Da Silva No. 146 e terreno junto a...                 Macao   
3                                2999 State Route 9L           Lake George   
4                           Lugar de Turbisquedo s/n           Turbisquedo   

                state countrycode       zip        lat         lng  status  \
0  Castilla-La Mancha         ESP     45600  39.963929   -4.819594       1   
1         East Sussex         GBR  BN20 7AJ  50.760040    0.281270       1   
2               Hunan         CHN      None  22.152700  113.557000       1   
3                  NY         USA     12845  43.441875  -73.649711       1   
4             Galicia         ESP     32151  42.482576   -7.763914       1   

      propertytype  rating  larbarterm  reviewscount  
0            Hotel     3.0         NaN           NaN  
1      Guest house     4.0         NaN           NaN  
2              Inn     2.0         NaN           NaN  
3            Hotel     3.0         NaN           NaN  
4  Bed & Breakfast     3.0         NaN           NaN

In [5]:
irl_hotel_details=hotel_details[hotel_details["countrycode"]=="IRL"]
irl_hotel_details.head(5)

id  hotelcode   hotelid  websitecode                   hotelname  \
155  4492621     459664  15690635            1  Auburndale Bed & Breakfast   
570  4493300     959939  46310864            1     Hillside Haven cottages   
773  4493646    1212015  26991076            1          Ocean View Doogort   
834  4493805    1294694  40029197            1             Rose of the Sea   
910  4493931    1350340    176109            1                    The Mont   

                         address          city             state countrycode  \
155  Smithland East, Springmount      Kilkenny   County Kilkenny         IRL   
570        82 1, Jayawardanapura  Nuwara Eliya  Central Province         IRL   
773            Ocean view Dugort       Doogort          Connacht         IRL   
834                Corclogh East     Belmullet          Connacht         IRL   
910               Merrion Square        Dublin          Leinster         IRL   

      zip        lat        lng  status  propertytype  rating  larbarterm  \
155  None  52.634443  -7.251345       1   Guest house     4.0         NaN   
570  None   6.923745  80.792331       1  Holiday home     3.0         NaN   
773  None  54.013329 -10.017120       0  Holiday Home     3.0         NaN   
834  None  54.205300  -9.963980       1      Homestay     0.0         NaN   
910  None  53.341477  -6.250687       1         Hotel     4.0         NaN   

     reviewscount  
155           NaN  
570           NaN  
773           NaN  
834           NaN  
910           NaN

In [6]:
hotel_rates=pd.read_parquet("/home/ai/Documents/my files/assignment_2/Pyspark_Assignment_2/raw_data_in/hospitality_raw_rates_20240922_123949.parquet")
hotel_rates.head(5)

websitecode         dtcollected  los  guests               roomtype  \
0            2 2024-09-22 18:19:47    1       1                   None   
1            2 2024-09-22 18:19:47    1       1  One-Bedroom Apartment   
2            2 2024-09-22 18:19:47    1       1  One-Bedroom Apartment   
3            2 2024-09-22 18:19:47    1       1  One-Bedroom Apartment   
4            2 2024-09-22 18:19:47    1       1  One-Bedroom Apartment   

   onsiterate  netrate currency  \
0         NaN      NaN      USD   
1        70.0     78.0      USD   
2        63.0      NaN      USD   
3        70.0      NaN      USD   
4        70.0      NaN      USD   

                                     roomdescription  \
0                                                nan   
1  Bedroom 1: 1 full bed , Living room: 1 sofa be...   
2  Bedroom 1: 1 bunk bed , Living room: 1 sofa be...   
3  Bedroom 1: 1 bunk bed , Living room: 1 sofa be...   
4  Bedroom 1: 1 bunk bed , Living room: 1 sofa be...   

                                    rateplan  ... pos  \
0                                        nan  ...   0   
1   Free cancellation before October 8, 2024  ...   0   
2                                          •  ...   0   
3  Free cancellation before October 12, 2024  ...   0   
4                                          •  ...   0   

                                           proxyused  israteperstay mealplan  \
0  brd-customer-hl_6d04aa29-zone-hotelsync_us-ip-...            nan      nan   
1  brd-customer-hl_6d04aa29-zone-hotelsync_us-ip-...              Y      nan   
2  brd-customer-hl_6d04aa29-zone-hotelsync_us-ip-...              Y      nan   
3  brd-customer-hl_6d04aa29-zone-hotelsync_us-ip-...              Y      nan   
4  brd-customer-hl_6d04aa29-zone-hotelsync_us-ip-...              Y      nan   

  optionalextras payplan id    refid   hotelid  minstay  
0            NaN     NaN  1   262438   4157796      NaN  
1            NaN     NaN  2  2055755   6578136      1.0  
2            NaN     NaN  3  1972651  10804142      1.0  
3            NaN     NaN  4  1972651  10804142      1.0  
4            NaN     NaN  5  1972651  10804142      1.0  

[5 rows x 33 columns]

JOININIG TWO DATAFRAMES WITH ID FIELD

In [7]:
data_merged=pd.merge(irl_hotel_details, hotel_rates, on="id", how="inner", suffixes=('_details', '_rates'))


In [ ]:
data_merged.head(5)

id  hotelcode hotelid_details  websitecode_details  \
0  4492621     459664        15690635                    1   
1  4493300     959939        46310864                    1   
2  4493646    1212015        26991076                    1   
3  4493805    1294694        40029197                    1   
4  4493931    1350340          176109                    1   

                    hotelname                      address          city  \
0  Auburndale Bed & Breakfast  Smithland East, Springmount      Kilkenny   
1     Hillside Haven cottages        82 1, Jayawardanapura  Nuwara Eliya   
2          Ocean View Doogort            Ocean view Dugort       Doogort   
3             Rose of the Sea                Corclogh East     Belmullet   
4                    The Mont               Merrion Square        Dublin   

              state countrycode   zip  ...  taxamount  pos  \
0   County Kilkenny         IRL  None  ...        NaN    0   
1  Central Province         IRL  None  ...        NaN    0   
2          Connacht         IRL  None  ...        NaN    0   
3          Connacht         IRL  None  ...        NaN    0   
4          Leinster         IRL  None  ...        NaN    0   

                                           proxyused israteperstay  \
0  brd-customer-hl_6d04aa29-zone-hotelsync_us-ip-...             Y   
1  brd-customer-hl_6d04aa29-zone-hotelsync_us-ip-...             Y   
2  brd-customer-hl_6d04aa29-zone-hotelsync_us-ip-...           nan   
3  brd-customer-hl_6d04aa29-zone-hotelsync_us-ip-...           nan   
4  brd-customer-hl_6d04aa29-zone-hotelsync_us-ip-...           nan   

                  mealplan  optionalextras  payplan    refid hotelid_rates  \
0                      nan             NaN      NaN  1668537      10667577   
1  Breakfast $0 (optional)             NaN      NaN  1388685       9706247   
2                      nan             NaN      NaN   178478      10622458   
3                      nan             NaN      NaN  1207925       3064775   
4                      nan             NaN      NaN   573278       3597481   

   minstay  
0      1.0  
1      1.0  
2      NaN  
3      NaN  
4      NaN  

[5 rows x 49 columns]

The country code of "IRELAND" is "IRL"

In [9]:
ireland_df=data_merged[data_merged["countrycode"]=="IRL"]


In [ ]:
ireland_df.head(5)

id  hotelcode hotelid_details  websitecode_details  \
0  4492621     459664        15690635                    1   
1  4493300     959939        46310864                    1   
2  4493646    1212015        26991076                    1   
3  4493805    1294694        40029197                    1   
4  4493931    1350340          176109                    1   

                    hotelname                      address          city  \
0  Auburndale Bed & Breakfast  Smithland East, Springmount      Kilkenny   
1     Hillside Haven cottages        82 1, Jayawardanapura  Nuwara Eliya   
2          Ocean View Doogort            Ocean view Dugort       Doogort   
3             Rose of the Sea                Corclogh East     Belmullet   
4                    The Mont               Merrion Square        Dublin   

              state countrycode   zip  ...  taxamount  pos  \
0   County Kilkenny         IRL  None  ...        NaN    0   
1  Central Province         IRL  None  ...        NaN    0   
2          Connacht         IRL  None  ...        NaN    0   
3          Connacht         IRL  None  ...        NaN    0   
4          Leinster         IRL  None  ...        NaN    0   

                                           proxyused israteperstay  \
0  brd-customer-hl_6d04aa29-zone-hotelsync_us-ip-...             Y   
1  brd-customer-hl_6d04aa29-zone-hotelsync_us-ip-...             Y   
2  brd-customer-hl_6d04aa29-zone-hotelsync_us-ip-...           nan   
3  brd-customer-hl_6d04aa29-zone-hotelsync_us-ip-...           nan   
4  brd-customer-hl_6d04aa29-zone-hotelsync_us-ip-...           nan   

                  mealplan  optionalextras  payplan    refid hotelid_rates  \
0                      nan             NaN      NaN  1668537      10667577   
1  Breakfast $0 (optional)             NaN      NaN  1388685       9706247   
2                      nan             NaN      NaN   178478      10622458   
3                      nan             NaN      NaN  1207925       3064775   
4                      nan             NaN      NaN   573278       3597481   

   minstay  
0      1.0  
1      1.0  
2      NaN  
3      NaN  
4      NaN  

[5 rows x 49 columns]

The basic overview of the dataset.

There are 31165 records of the hotels in ireland with its rates.

The datatypes consists of object int and float .

The column "Larbarterm" is kept optional and contains no record. So it is good to drop that column, inorder to make data effective to analyse.

The column "reviewscount" have only few records and it is also good to drop the column.

The column zip also contains a lot of null values and it is also good to drop the column

The room type and minstay column ahs nearly 4803 record as null which is about 15.4% of the records  lets try to impute the values 



Since this is a rating column people choose the hotels mostly based on rating in a booking app.So this is an important column to be considered.

In [ ]:
print("Shape of the dataset:", ireland_df.shape)
# print("\nData types:\n", ireland_df.dtypes)
print("\nMissing values:\n", ireland_df.isnull().sum())

Shape of the dataset: (31165, 49)

Missing values:
 id                         0
hotelcode                  0
hotelid_details            0
websitecode_details        0
hotelname                  0
address                   49
city                       4
state                    638
countrycode                0
zip                    30822
lat                        0
lng                        0
status                     0
propertytype               0
rating                    48
larbarterm             31165
reviewscount           31008
websitecode_rates          0
dtcollected                0
los                        0
guests                     0
roomtype                4803
onsiterate              4803
netrate                26849
currency                   0
roomdescription            0
rateplan                   0
sourceurl                  0
roomamenities              0
occupancy                  0
ispromo                    0
closed                     0
checkin             

In [12]:
# ireland_df.drop(["zip","larbarterm","reviewscount","onsiterate","netrate"],axis=1,inplace=True)

We can see that in the columns of mealplan and israteperstay has null values shows as nan which is the dataframe would consider it has a string and says it has values but it is actually a null so we need to replace it as null values and decide whether it is good to impute or drop 

In [13]:
print(ireland_df['mealplan'].unique())

# Check unique values in the 'israteperstay' column
print(ireland_df['israteperstay'].unique())

['nan' 'Breakfast $0 (optional)' 'Breakfast $3.25 (optional)'
 'Very good breakfast included' 'Good breakfast $6'
 'Good breakfast included' 'Continental breakfast included'
 'Breakfast included in the price' 'Excellent breakfast included'
 'Buffet breakfast included' 'Breakfast & dinner included'
 'Wonderful breakfast $6' 'Breakfast $3 (optional)'
 'Breakfast $7 (optional)' 'Breakfast $2 (optional)'
 'Wonderful breakfast included' 'Very good breakfast $20'
 'Wonderful breakfast $23' 'Breakfast, lunch & dinner included'
 'Exceptional breakfast included' 'Excellent breakfast $14'
 'Dinner $10 (optional)' 'Good breakfast $22' 'Very good breakfast $12'
 'Wonderful breakfast $14' 'Very good breakfast $2' 'Good breakfast $12'
 'Exceptional breakfast $39' 'Exceptional breakfast $13'
 'Breakfast $5 (optional)' 'Good breakfast $9' 'Good breakfast $11'
 'Breakfast $27 (optional)' 'Breakfast $8 (optional)'
 'Very good breakfast $27' 'Exceptional breakfast $6' 'Good breakfast $18'
 'Very good bre

replacing the columns

In [14]:
# import numpy as np

# Replace 'nan' strings with actual NaN values
ireland_df.replace('nan', np.nan, inplace=True)

# Check for null values again
print(ireland_df[['mealplan', 'israteperstay']].isnull().sum())

mealplan         16057
israteperstay     4803
dtype: int64


PERFORMING THE EDA(EXPLORATORY DATA ANALYSIS) PROCESs USING SWEETVIZ

In [15]:
# Get descriptive statistics for numerical columns
print(ireland_df.describe())

# Get descriptive statistics for categorical columns
print(ireland_df.describe(include=['object']))


                 id     hotelcode  websitecode_details           lat  \
count  3.116500e+04  3.116500e+04         31165.000000  31165.000000   
mean   4.157941e+06  1.579156e+06           935.944970     53.116141   
std    2.423786e+06  1.348267e+06          1545.953455      1.405095   
min    5.720000e+02  3.210000e+02             1.000000    -33.963164   
25%    2.068724e+06  4.584440e+05             2.000000     52.350100   
50%    4.147855e+06  1.264023e+06            34.000000     53.280200   
75%    6.248012e+06  2.513599e+06           324.000000     53.555490   
max    8.297229e+06  5.668045e+06          3744.000000     57.973383   

                lng        status        rating  larbarterm  reviewscount  \
count  31165.000000  31165.000000  31117.000000         0.0    157.000000   
mean      -7.237807      0.788224      2.609538         NaN    581.649682   
std        4.209028      0.408573      1.587816         NaN   1552.094131   
min     -124.012026      0.000000      0.00

USING SWEETVIZ LIBRARY TO ANALYZE AND MAKE REPORTS


In [17]:
import sweetviz as sv
from PIL import Image


# Create a Sweetviz report
report = sv.analyze(ireland_df)

# Show the report
report.show_html('ireland_df_report.html')


/home/ai/.local/lib/python3.10/site-packages/sweetviz/dataframe_report.py:75: FutureWarning: iteritems is deprecated and will be removed in a future version. Use .items instead.
  all_source_names = [cur_name for cur_name, cur_series in source_df.iteritems()]
/home/ai/.local/lib/python3.10/site-packages/sweetviz/dataframe_report.py:110: FutureWarning: iteritems is deprecated and will be removed in a future version. Use .items instead.
  filtered_series_names_in_source = [cur_name for cur_name, cur_series in source_df.iteritems()


                                             |          | [  0%]   00:00 -> (? left)

/home/ai/.local/lib/python3.10/site-packages/sweetviz/series_analyzer_numeric.py:25: FutureWarning: The 'mad' method is deprecated and will be removed in a future version. To compute the same result, you may do `(df - df.mean()).abs().mean()`.
  stats["mad"] = series.mad()
/home/ai/.local/lib/python3.10/site-packages/sweetviz/series_analyzer_numeric.py:25: FutureWarning: The 'mad' method is deprecated and will be removed in a future version. To compute the same result, you may do `(df - df.mean()).abs().mean()`.
  stats["mad"] = series.mad()
/home/ai/.local/lib/python3.10/site-packages/sweetviz/series_analyzer_text.py:19: FutureWarning: iteritems is deprecated and will be removed in a future version. Use .items instead.
  for item in to_process.source_counts["value_counts_without_nan"].iteritems():
/home/ai/.local/lib/python3.10/site-packages/sweetviz/series_analyzer_numeric.py:25: FutureWarning: The 'mad' method is deprecated and will be removed in a future version. To compute the sam

Report ireland_df_report.html was generated! NOTEBOOK/COLAB USERS: the web browser MAY not pop up, regardless, the report IS saved in your notebook/colab files.


DESCRIBING EACH FIELD WITH ITS PREPROCESSING AND STANDARDIZATION PROCESS.


ID:["This is unique variable in both the dataframes","It hepls to join the two dataframes get out the records of ireland","it doesn't have any null values"," This is a primary contraints which has highest correlation only with the proxyuse not with any other column"]
PREPROCESSING: not required  already
STANDARDIZATION: Already consists unique values


hotelcode:["It contains only 8470 unique values which is about 27%","There is no missing values either","it is srtongly correlated with propertype with the correlation rate of 0.59","next is rating and the next will be status".]
PREPROCESSING: not required  already
STANDARDIZATION: Already consists unique values

hotelid: ["this column basically contains the id of the hotel. there is no missing values and it contains 21176 distinct features.]
PREPROCESSING: not required  already
STANDARDIZATION: Already consists unique values

websitecode: [" the website code column has no null values and it is highly correlated with status and propertytype with 0.31 and 0.14.]
PREPROCESSING: not required  already
STANDARDIZATION: Already consists unique values

hotel name: ["this column especially talks about the name of the hotel in the ireland]" it contains only 27% unique names and there is no null values"]
PREPROCESSING: convert the name into lower case. removal of special character.
STANDARDIZATION:normalizing the hotel name into standard format 

Address : [" this address column states the where the particular hotel is located and it contains 49 missing values which is nearly 1%.it contains 7825 unique values out of 31116.since there are lot of hotels names are repeated probably the address also."]
PREPROCESSING: Handle the missing values may be we can impute the values since there is only 49. else we can replace the null values with "unknown value"
STANDARDIZATION:normalizing it to the standard format

city: ["In this column the it contains only 1336 unique values where the most of the hotels located in one or more cities", it also contains 4 null values .]
PREPROCESSING: Handling the missing values by imputing.
STANDARDIZATION:normalizing it to the standard format

state: ["In the state column we have about 98% records. Additionaly we have 105 distinct records"]
PREPROCESSING: There is no missing values and preprocessing includes removal of punctuation marks
STANDARDIZATION: normalize to standard format.

countrycode: [ to talk about country code we have filtered only one country that is ireland we are analysing the records in ireland so there is no null values.]
PREPROCESSING: not required
STANDARDIZATION: not required

zip code :["It is a feature used in foriegn countries to deliver the mail faster.most probaly to find out the location",but this column has 99% of null values and there are only 343 values out of which 297 is unique. There are 139 postal codes in ireland.]
PREPROCESSING: Since there is a lot of missing values we could drop the column else we could map with the cities and impute the null values accordingly.Else it would become a mess.
STANDARDIZATION:There is particular format for zip code according to their cities. Normalizing is required.

latitude: [this latitude region will show only the latitude of the region in ireland beacuse we havefiltered the records only for the ireland.ther is no null values as well and there are 7925 unique features.]
PREPROCESSING: not required 
STANDARDIZATION: Already consists unique values


longitude:  [this is similar to latitude but it contains 8521 unique values.]
PREPROCESSING: not required 
STANDARDIZATION: Already consists unique values


status:[this columns shows whether these hotels are active or not by indicating 1 to show it is active and 0 to show it is inactive]
PREPROCESSING: not required 
STANDARDIZATION: It is already standardized.


propertytype : this column tell about wether the property is a hotel or apartments or hostel or guest house and so on. basically it speaks about kind of property. There are 52 unique features. Specifically this column provides the informations countrycode,websitecode_rates,los,guests,currency,checkin and checkout and pos with the value of 1.
PREPROCESSING: There is no missing values and removal of punction marks are necessary
STANDARDIZATION: Should be with in a range.


rating : the rating column has only 10 uniques values which ranges from 1-5 with the interval of 0.5 . This ia an important feature for the hospitality because people choose property according to their ratings. There is no rating for about 48 records.
PREPROCESSING: Here handling missing values is required, but then we cannot give rating based on mode that would be inappropriate. We could handle missing values using mean or median that might work.


larberterm : This column has no values so, it is difficult to know what actually larbarterm refers to.
PREPROCESSING: This field is not required better you can drop the column.
STANDARDIZATION: Not required.

reviewscount : This column basically means the counts of the review  and it also contains 99% of the null values.
PREPROCESSING:Better to drop the column
STANDARDIZATION: Not required.


website rates : which is basically contains only 1 value due to the filteration of ireland data.
PREPROCSSING: Not required
STANDARDIZATION: Not required.

dtcollected: This might be the dates that is the data collected on. so there are 9570 unique records and does not contain any null value.
PREPROCESSING:Convert to a standard date format (e.g., YYYY-MM-DD)
STANDARDIZATION: Not required


los and guests contain only one value  as 1 and i dont find something important. this value is based on the fileteration may be.
PREPROCESSING: Not Required
STANDARDIZATION; Not required

room type : This is column talks about the type of rooms that the hotels have might be 1BHK, 2BHK, 3BHK  etc.,There are missing 4803 values  and with the 3518 uniue values.
PREPROCESSING: Handle the missing values by imputing.
STANDARDIZATION:Covert to the standard format.

on site rate: this speaks about the rate of the site whcih have 1140 distinct values and 4803 missing values .
PREPROCESING: The missing values should be handled by imputing the values.
STANDARDIZATION: Can be normalized to a standard scale(eg.1-100)

net rate: net rate has a huge impact on the hotels growth.This shows 26849 null values  which is a huge and there are only 632 unique records.
PREPROCESSING:There is a huge null values. It cannot be imputed as well.
STANDARDIZATION:Can be normalize to a standard scale(eg 0-100)

currency : currency is basicaly here represents the currency used in ireland. So there will be only one unique record.
PREPROCESSING: Not Required
STANDARDIZATION: Not Required

room description: This column describes the room in each hotel like king bed, queen bed, twin beds with its size etc.,This contains about 7872 unique features and contains 4889 null values which is about 16% 
PREPROCESSING: Handling missing values is required.
STANDARDIZATION: Convert into stanadard format withou and extra spaces

rate plan : In rate plan there are features like free cancallation and refundable non-refundable etc. Like wise there are 75 distinct features and 4837 missing values.
PREPROCESSING:Should handle the missing values. Removal of punctuation marks is required.
STANDARDIZATION: convert to a standard format


source url : this column is basically where the data is fetched from. There are 1712 distinct sources but most of the records are fetched from booking.com. these 1712 values are one of the pages in the booking.com
PREPROCESSING: we could handle the missing value by imputing it.
STANDARDIZATION: confusing

room amenities: This is clear that it talks about the facilities in the hotels or apartments or any kinda property. THe features are like free wifi , tv, refrigerator, ironing facilities, laundary and so on. There are 15152 distinct features.and 4902 null values
PREPROCESSING: Handle the missing values by imputing but however it wouldn't be fine. The punctuation and stop words has to be removed.

occupany : this column is about how many people can be accomodated within that property.and it has 36 unique values and there is no missing values.This field is import to select the hotels because we can arrange the room accordingly.
PREPROCESSING:Not Required 
STANDARDIZATION: wihtin a scale of 1-4


ispromo: I guess it might talk about the promocode for the discounts . there are only two unique values . one is y which stands for yes and N whcih stands for No.
PREPROCESSING: not required
STANDARDIZATION: not required.

closed: This columns talks about whether the property is open or closed. if it opened it is Y and closed it is N.After the keen observation the count of N is high compare to Y in both ispromo and closed columns.
PREPROCESSING: Not required.
STANDARDIZATION: Not required

checkin : This data is collected on the date 13.10.2024 so there is only one value in this column.These records are based on the checkin of the people on the date of 13.10.2024.
PREPROCESSING: Not Required.
STANDARDIZATION: Not required


checkout : This datais also collected on the date 14.10.2024 . these data is based on people who checkout on 14.10.2024.
PREPROCESSING: Not Required.
STANDARDIZATION: Not required


discount : It speaks about whether there is any discount for the people who are booking on the particular date. this column contains 26849 null values and 1399 unique values.
PREPROCESSING: need to handle the null value by providing it as 0. Because we cannot impute the null values using central tendency, it would not be appropriate.
STANDARDIZATION: Not Required already in a range of 1-100

promo name: It tells about the promo code like new propert deal , last escape deal and so on.it also contains null values 
PREPROCESSING: These promo name is a unique contraints and the missing values has to imputed by maping with discount or amenities.
STANDARDIZATION: Convert it into a standard format

status code : It usually speaks about whether the promo are active . there is only 2 distinct values 
PREPROCESSING:not required.
STANDARDIZATION: not required.

taxstatus : It is all about tax for th no.of people staying in the hotels there is only 4 unique values and no missing records.
PREPROCESSING: not required.
STANDARDIZATION: Should  set a scale limit.

taxtype : amount plus, taxes and fee  for staying only fee and taxes. may be extra charges. Containd missing values
PREPROCESSING: While encoding we could make a label encoder.should handle missing values
STANDARDIZATION: not required.

taxamount: the tax amount is mostly between 10-20 and most of the taxes are between the earlier mentioned range and few are between 20 -50 and rarely above 60. there are (99%) records missing.
PREPROCESSING : better this can be drioped else we could make some calculation and fix it


pos: I don't understand about this pos might be it talks about the position  and there is only one values due to the filteration.
PREPROCESSING: Not Required.
STANDARDIZATION: Not Required.

proxy used: There are nearly 100 unique proxy used to get the data from the websites and does not contain any null value.
PREPROCESSING: confused.

is rate per stay= This column shows that whether there is rate included per stay. There is only on value that is Y. It also contains 4803 null values.
PREPROCESSING: The missing values has to be handled.
STANDARDIZATION: not required.

meal plan: This is important feature to note because if there is anything like no meal provided poeple need to search for the restaurant to have their meal. It shows that most of hotels provide breaksfast as complementary. let us anlyze more. It also contains 16057 null values and 379 disctinct values.
PREPROCESSING: IT has to be preprocessed in such a way that if breakfast is provided it should be encoded as 1 .. likewise for other.
STANDARDIZATION : not required if it is encode properly.

optional extras: This has NO value and it is difficult to find what is signifies.
PREPROCESSING: It is better to drop the column.
STANDARDIZATION: not required.

payplan : this also the same. there is no plan given for any of the records.
PREPROCESSING; Good to drop the column.
STANDARDIZATION:not required.

ref id : ref id signifies that there is any referal id for a paricular stay or property.There is no missing values.
PREPROCESSING: may be this could be mapped with the promo code or discount and make some calculations out of it.
STANDARDIZATION; not required.

hotelid rates: this column has a large no seems to be the rate of the hotel id. and there is also no missing values.
PREPROCESSING: confused 

min stay : this column has only one value as 1. most probably seems to be minimum no.of people can stay. This contains 4803 null values.
PREPROCESSING: missing values has to be handled or replace with ones.

ALL THE CATEGORICAL COLUMN HAS TO BE ENCODED.